numpyro/distributions/continuous.py

```diff
@@ -1033,6 +1033,115 @@ def icdf(self, q):
+            scale_rows (_type_): _description_
+            validate_args (_type_, optional): _description_. Defaults to None.
+
+        Raises:
+            ValueError: _description_
+        """
+        if not (loc.ndim == scale_columns.ndim == scale_rows.ndim):
+            raise ValueError(
+                "ndim of loc, scale_columns and scale_rows are reuqired to be equal."
+            )
+
+        event_shape = loc.shape[-2:]
+        n, p = event_shape
+        batch_shape = loc.shape[:-2]
+
+        assert scale_rows.shape == batch_shape + (
+            n,
+            n,
+        ), "scale_rows.shape does not match"
+        assert scale_columns.shape == batch_shape + (
+            p,
+            p,
+        ), "scale_rows.shape does not match"
+
+        super(MatrixNormal, self).__init__(
+            batch_shape=batch_shape, event_shape=event_shape
+        )
+
+        self.loc = loc
+        self.scale_columns = scale_columns
+        self.scale_rows = scale_rows
+
+    @property
+    def mean(self):
+        return self.loc
+
+    def sample(self, key, sample_shape=()):
+
+        mvn_dim = np.prod(self.event_shape)
+        loc = jnp.reshape(self.loc, newshape=self.batch_shape + (mvn_dim,), order="F")
+        assert loc.shape == self.batch_shape + (mvn_dim,), "vec shape does not match"
+
+        cov = _kron(self.scale_rows, self.scale_columns)
+        assert cov.shape == self.batch_shape + (
+            mvn_dim,
+            mvn_dim,
+        ), "cov shape does not match"
+
+        mvn = MultivariateNormal(loc=loc, covariance_matrix=cov)
+
+        assert mvn.batch_shape == self.batch_shape, "batch shape of MVN does not match"
+        assert mvn.event_shape == (
+            np.prod(self.event_shape),
+        ), "event shape of MVN does not match"
+
+        samples_mvn = mvn.sample(key, sample_shape=sample_shape)
+        assert samples_mvn.shape == sample_shape + self.batch_shape + (
+            mvn_dim,
+        ), "MVN sample shape does not match"
+        samples_mn = samples_mvn.reshape(
+            sample_shape + self.batch_shape + self.event_shape, order="F"
+        )
+        assert (
+            samples_mn.shape == sample_shape + self.batch_shape + self.event_shape
+        ), "MN sample shape does not match"
+
+        return samples_mn
+
+    def log_prob(self, value):
+        mvn_dim = np.prod(self.event_shape)
+        loc = jnp.reshape(self.loc, newshape=self.batch_shape + (mvn_dim,), order="F")
+        assert loc.shape == self.batch_shape + (mvn_dim,)
+
+        cov = _kron(self.scale_rows, self.scale_coloumns)
+        assert cov.shape == self.batch_shape + (
+            mvn_dim,
+            mvn_dim,
+        ), "cov shape does not match"
+
+        mvn = MultivariateNormal(loc=loc, covariance_matrix=cov)
+        values = jnp.reshape(
+            self.loc, newshape=self.batch_shape + (mvn_dim,), order="F"
+        )
+        log_prob_mvn = mvn.log_prob(values)
```

Comment:
This seems to be expensive: creating a matrix (loc.shape[-1] * loc.shape[-2], loc.shape[-1] * loc.shape[-2]). I think we can use the formula in [wikipedia](https://en.wikipedia.org/wiki/Matrix_normal_distribution#Definition)
```
row_log_det = jnp.log(jnp.diagonal(self.scale_tril_row, axis1=-2, axis2=-1)).sum(-1)
col_log_det = jnp.log(jnp.diagonal(self.scale_tril_column, axis1=-2, axis2=-1)).sum(-1)
# compute the trace term
diff = vale - self.loc
diff_row_solve = solve_triangular(self.scale_tril_row, diff, lower=True)
diff_col_solve = solve_triangular(self.scale_tril_column, jnp.swapaxes(diff_row_solve, -2, -1), lower=True)
trace_term = jnp.square(diff_col_solve.reshape(diff_col_solve.shape[:-2] + (-1))).sum(-1)
```

Reply:
It is. Your suggestion is much better. But I do not fully understand the calculation of the`trace_term`.
Using the notation from Wikipedia and `D=(X-M)` the `trace_term` is given by `trace_term=tr(inv(V)@D.T@inv(U)@D)`. Since we us a lower triangular parametrisation`ìnv(U)@D=inv(scale_row_tril@scale_row_tril.T)@D`=inv(scale_row_tril.T)@inv(scale_row_tril)@D=solve_triangular(lower_row_tril,solve_triangular(scale_row_tril,D,lower=True),lower=False)`

Reply:
We can rewrite `tr(inv(V)@D.T@inv(U)@D)` into
```
tr(inv(V_tril).T@inv(V_tril)@D.T@inv(U_tril).T@inv(U_tril)@D)
= tr(inv(V_tril)@D.T@inv(U_tril).T @ inv(U_tril)@D@inv(V_tril).T)
= tr(X.T @ X)
where X = inv(U_tril) @ (inv(V_tril) @ D.T).T
```

Reply:
I've only looked at the term inside the trace but the key is the cyclic property of the trace (`tr(A@B@C)=tr(C@A@B)`) right?

Reply:
That's right. That helps us factorize the trace term into the form X @ X.T, so that we can compute trace easily.

Reply:
The current implementation works for single samples. With `vmap` I also was able to e broadcasted to `1-D`samples. For other cases I need a little more time.

Reply:
@fehiepsi I've added a draft for a `log_prob`method which should deal with arbitrary sample and batch shapes.